06-classification_continued.Rmd

```diff
@@ -559,16 +559,257 @@ The overall workflow for performing K-nearest neighbour classification using `ti
+                summarize(n = n()) %>%
+                mutate(frac = n/sum(n)) %>%
+                summarize(mx = max(frac)) %>%
+                select(mx)
+base_acc <- base_acc$mx %>% unlist()
+
+# plot
+res <- tibble(ks = ks, accs = accs)
+res <- res %>% mutate(base_acc = base_acc)
+plt_irrelevant_accuracies <- res %>%
+  ggplot() +
+  geom_line(mapping = aes(x=ks, y=accs, linetype="Tuned K-NN")) +
+  geom_hline(data=res, mapping=aes(yintercept=base_acc, linetype="Always Predict Benign (63%)")) +
+  labs(x = "Number of Irrelevant Predictors", y = "Model Accuracy Estimate") + 
+  scale_linetype_manual(name="Method", values = c("dashed", "solid"))
+
+plt_irrelevant_accuracies
+```
+</center>
+
+Wow! Once we add sufficiently many irrelevant predictor variables, our classifier
+does not do much better than just always guessing "benign", 
+ the most common class label in the data (63% of observations are benign).
+
+So then if we can't just throw in the kitchen sink, how do we choose which
+variables to use as predictors?  This is a very difficult question to answer in
+general, and there are a number of methods that have been developed that apply
+in particular cases of interest; see the additional resources at the end of
+this chapter to find out where you can learn more about variable selection.
+Here we will discuss *best subset selection,* which is applicable to any 
+classification method (K-NN or otherwise), and works when you have a small
+number of total variables from which you want to keep only the relevant ones (to
+avoid the problem in the example above). Best subset selection works 
+as follows. If you have $m$ predictors, you: 
+
+1. create a separate model for every possible subset of predictors,
+2. tune each one using cross validation, and 
+3. finally pick the subset of predictors that gives you the highest estimated prediction accuracy.  
+
+Note that best subset selection becomes very slow when you have even a moderate
+number of predictors to choose from. This is because the number of possible predictor subsets
+grows very quickly with the number of predictors. For example, if we have $m=2$ predictors---let's call
+them A and B---then we have 3 variable sets to try: A alone, B alone, and finally A
+and B together. If we have $m=3$ predictors---A, B, and C---then we have 7
+to try: A, B, C, AB, BC, AC, and ABC. In general, the number of models
+we have to train for $m$ predictors is $2^m-1$; in other words, when we 
+get to $m=10$ predictors we have over *one thousand* models to train, and 
+at $m=20$ predictors we have
+over *one million* models to train! If you have more than about 10 predictors,
+you should first use your scientific understanding of the data
+to choose the 10 or so predictors that you think are most meaningful, and then
+run best subset selection within those 10. Alternatively, you can use a different
+predictor variable selection method; see the additional resources at the end
+of this chapter. 
```

Comment:
OK, I see here you define what small vs large are. So best subset will not work for your 502 example above likely. 

Also, what about number of observations... that will impact this as well, but not in the same way. Should we mention that?

Reply:
edited

(I removed explicit reference of the 500 variables here)